### PR TITLE
[SC-452] Refactor forwarders and receivers

### PR DESCRIPTION
**Pull request by hexonaut**
This PR does two refactors:

1. Split out `XChainForwarders` into separate libraries for better organization. I've also added some missing L2 -> L1 messages on each forwarder.
2. Receivers don't need to be extended, as the logic will always proxy a function call such as here: https://github.com/marsfoundation/spark-gov-relay/blob/master/src/receivers/BridgeExecutorReceiverArbitrum.sol#L19 . This update makes receivers generic message-passers.

You can see concrete example of how these changes improve usage here: https://github.com/marsfoundation/spark-gov-relay/pull/18/files and here: https://github.com/marsfoundation/xchain-dsr-oracle/pull/20/files

### CODE REVIEW

File: test/CircleCCTPIntegration.t.sol [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    using CCTPBridgeTesting for *;
+    using DomainHelpers     for *;
+
+    uint32 sourceDomainId = CCTPForwarder.DOMAIN_ID_CIRCLE_ETHEREUM;
+    uint32 destinationDomainId;
+
+    // Use Optimism for failure tests as the code logic is the same
+
+    function test_invalidSourceAuthority() public {
```
Comment (lucas-manuel) [inline test/CircleCCTPIntegration.t.sol:20]: This one should be third

File: test/OptimismIntegration.t.sol [hunk 2]
```diff
@@ -24,90 +14,74 @@ contract OptimismIntegrationTest is IntegrationBaseTest {
...
     event FailedRelayedMessage(bytes32);
 
-    function test_optimism() public {
-        checkOptimismStyle(getChain("optimism").createFork());
-    }
+    // Use Arbitrum One for failure test as the code logic is the same
 
-    function test_base() public {
-        checkOptimismStyle(getChain("base").createFork());
+    function test_invalidSourceAuthority() public {
```
Comment (lucas-manuel) [inline test/OptimismIntegration.t.sol:19]: This should be second

File: test/OptimismIntegration.t.sol [hunk 3]
```diff
@@ -24,90 +14,74 @@ contract OptimismIntegrationTest is IntegrationBaseTest {
 
     event FailedRelayedMessage(bytes32);
 
-    function test_optimism() public {
-        checkOptimismStyle(getChain("optimism").createFork());
-    }
+    // Use Arbitrum One for failure test as the code logic is the same
```
Comment (lucas-manuel) [inline test/OptimismIntegration.t.sol:17]: Update comment

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/receivers/ArbitrumReceiver.sol [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * @title  ArbitrumReceiver
+ * @notice Receive messages to an Arbitrum-style chain.
+ */
+contract ArbitrumReceiver {
+
+    using Address for address;
+
+    address public immutable l1Authority;
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/receivers/ArbitrumReceiver.sol:14]: I've thought about this some, and I think it's more clear to use l1/l2 in the case of native bridges. They are directional, and it makes it clear the direction. Source/Destination can be either direction, but this receiver is directioned.

File: src/receivers/OptimismReceiver.sol [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * @title  OptimismReceiver
+ * @notice Receive messages to an Optimism-style chain.
+ */
+contract OptimismReceiver {
+
+    using Address for address;
+
+    ICrossDomainOptimism public constant l2CrossDomain = ICrossDomainOptimism(0x4200000000000000000000000000000000000007);
+
+    address public immutable l1Authority;
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/receivers/OptimismReceiver.sol:20]: Same here. Want to keep it more specific to the native bridge language.

File: src/receivers/OptimismReceiver.sol [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * @title  OptimismReceiver
+ * @notice Receive messages to an Optimism-style chain.
+ */
+contract OptimismReceiver {
+
+    using Address for address;
+
+    ICrossDomainOptimism public constant l2CrossDomain = ICrossDomainOptimism(0x4200000000000000000000000000000000000007);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/receivers/OptimismReceiver.sol:18]: Same.

File: src/receivers/AMBReceiver.sol [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * @title  AMBReceiver
+ * @notice Receive messages to AMB-style chain.
+ */
+contract AMBReceiver {
+
+    using Address for address;
+
+    address public immutable amb;
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/receivers/AMBReceiver.sol:19]: Same.

File: src/receivers/AMBReceiver.sol [hunk 8]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import { Address } from "openzeppelin-contracts/contracts/utils/Address.sol";
+
+interface IArbitraryMessagingBridge {
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/receivers/AMBReceiver.sol:6]: Going to leave as-is then.

File: README.md [hunk 9]
```diff
@@ -1,5 +1,20 @@
 # xchain-helpers
 
-This repository contains tooling for multichain testing such as cross-chain e2e testing support.
+This repository three tools for use with multi-chain development. Domains refer to blockchains which are connected by bridges. Domains may have multiple bridges connecting them, for example both the Optimism Native Bridge and Circle CCTP connect Ethereum and Optimism domains.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply README.md:3]: fixed

File: README.md [hunk 10]
```diff
@@ -1,5 +1,20 @@
...
+
+## Forwarders
+
+These libraries provide standardized syntax for sending a message to a bridge.
+
+## Receivers
+
+The most common pattern is to have an authorized contract forward a message to another "business logic" contract to abstract away bridge dependencies. Receivers are contracts which perform this generic translation - decoding the bridge-specific message and forwarding to another `target` contract. The `target` contract should have logic to restrict who can call it and permission this to one or more bridge receivers.
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply README.md:13]: added another paragraph

File: test/CircleCCTPIntegration.t.sol (hunk 1 above)
Comment (hexonaut) [reply test/CircleCCTPIntegration.t.sol:20]: fixed

File: test/OptimismIntegration.t.sol (hunk 3 above)
Comment (hexonaut) [reply test/OptimismIntegration.t.sol]: fixed

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**